File: PA01/train.jl
#= train.jl
training procedure for perceptron.jl
PA01 (CS5000)
01170065 - Xavier Sánchez Díaz
=#

include("perceptron.jl")

"""
Training procedure.
Trains the perceptron using a set of examples `P` and their expected outputs `T`, over `its` iterations. Returns a list of the form `[W, b]`.

`train(P, T, its)`
"""
function train(P, T, its)
    W = randn(size(T,2), size(P,2))
    b = randn(size(W,1), 1)
    i = 1

    for it in its
        p = P[i,:]
        t = T[i,:]
        a = perceptron(W, p, b)
        err = t - a
        W = W + err * p'
        b = b + err
        i += 1

        # start over
        if i > size(P,1)
            i = 1
        end
    end

    return [W, b]
end

File: PA02/train.m
# -- Mapping Function: train(P, T, maxIterations)
#     Trains a perceptron by using a set of examples represented 
#     by P (the inputs) and T (their expected outputs). 

function [W, b] = train(P, T, maxIterations)
	W = rand(columns(T), columns(P));
	b = rand(rows(W), 1);
	i = 1;	
	for x = 1:maxIterations
		p = P(i, :)';
		t = T(i, :)';
		a = perceptron(W, b, p, @hardlim);
		e = t - a;
		W = W + e * p';
		b = b + e;
		i = i + 1;
		if i > rows(P)
			i = 1;
		end
	end
end
